### ISSUE TITLE
3.0.0

### ISSUE THREAD

**Issue by vegarringdal** [fuse-box/fuse-box-typechecker#58]
Will try and add faster checking by using IncrementalCompilerHost

**Comment by vegarringdal** [comment]
If anyone want to try if its faster then use this
`fuse-box-typechecker@3.0.0-next.2`

**Comment by vegarringdal** [comment]
Some bads parts here, some todo for me for next week:
- [ ] remove some of the weird code
- [ ] move builtin watch into thread
- [ ] using tread to check code should also print result in tread so it dont interrupt main thread

**Comment by vegarringdal** [comment]
@nchanged 
When I update it I want it to be easier to use with fusebox.
I use it for otherstuff too, but since its called fusebox-typechekcer it might be nice if it was super easy to use.

Will the v4 plugin api have events for something like this:
- 1. first run
  - with current tsconfig
  - with current watch settings
- 2. Update/watch triggered
  - once when all is done
- 3. Abort- config changed- need to shut down

Maybe best if it just runs its own tread all the time and uses it own watch so it dont mess with fusebox speed. Atm I notice I need to improve some parts, they are just making it hard for fusebox 😂

**Comment by nchanged** [comment]
Here is a simple `v4` plugin

```
export function pluginTypeChecker(opts) {
  return (ctx: Context) => {
    ctx.ict.on('complete', props => {
      // initial run
      return props;
    });
    ctx.ict.on('rebundle_complete', props => {
      // re-bundle
      return props;
    });
  };
}

```

https://github.com/fuse-box/fuse-box/blob/26cc7f28b2d5c1a7ff12cee1f41827c4e9becaf4/src/interceptor/events.ts#L36-L43

**Comment by nchanged** [comment]
WatcherAction

**Comment by vegarringdal** [comment]
@nchanged 
thx

Just got old grid cleaned up and removed a lot
`fuse-box-typechecker@3.0.0-next.3``
Will try and look at the plugin after Ive tested new code a little more.
Maybe clean up a little more when Ive had some more sleep 😂 

New version have fewer options and more stuff it can do in the worker.
```ts
printSettings(): void;
inspectAndPrint(): number;
inspectOnly(oldProgram: ts.EmitAndSemanticDiagnosticsBuilderProgram): IResults;
printOnly(errors: IResults): number;
worker_watch(pathToWatch: string): void;
worker_kill(): void;
worker_inspect(): void;
worker_PrintSettings(): void;
worker_print(): void;
worker_inspectAndPrint(): void;
```

Im think this will be the sequenze

1.   starting -> worker_PrintSettings();
2.   on change -> worker_inspectAndPrint();
3.   if it needs to quit -> worker_kill()

Or we can just run the watch in the backgound so we dont use time to call it

**Comment by vegarringdal** [comment]
Cleaned up more
Tslint is removed since it will be deprecated some time in 2019
Added missing throwOn options
Removed emit part, since anyonecan do it with a since line..
`node ./node_modules/typescript/bin/tsc --build`

`fuse-box-typechecker@3.0.0-next.4`

Todo:
- [ ] Make plugin

**Comment by vegarringdal** [comment]
plugin added

`fuse-box-typechecker@3.0.0-next.5`

**Comment by vegarringdal** [comment]
Used no fusebox tsconfig if no path is set
You can still overide if you want

`fuse-box-typechecker@3.0.0-next.6`

**Comment by vegarringdal** [comment]
Todo 
when adding basepath I need to add this to file source, else contrl+mouse click will fail

fixed [ ]

**Comment by vegarringdal** [comment]
fuse-box-typechecker@3.0.0-next.10

fixed basepath issues when using server (2 plugins)
fixed production build, if `.NODE_ENV === "production"` then it does not use threads. (imprortant so it does not hold process alive after fusebox is done)

**Comment by vegarringdal** [comment]
fuse-box-typechecker@3.0.0-next.30

removed ref to homeDir

**Comment by vegarringdal** [comment]
released v 3.0.0
